Docs: How to run external codes (#4114)
aiidateam/aiida-core: docs/source/howto/codes.rst

Running external codes
======================

> To run an external code with AiiDA, you will need to use an appropriate :ref:`calculation plugin <topics:plugins>` that knows how to take transform the input nodes into the input files that the code expects, copy everything in the code's machine, run the calculation and retrieve the results.

Reviewer: ```suggestion
To run an external code with AiiDA, you will need to use an appropriate :ref:`calculation plugin <topics:plugins>` that knows how to transform the input nodes into the input files that the code expects, copy everything in the code's machine, run the calculation and retrieve the results.
```

This plugin would have already been created, either by you (see section on :ref:`interfacing external codes <how-to:codes:plugin>`) or by other AiiDA plugin-developers (available packages are listed in the `plugin registry <https://aiidateam.github.io/aiida-registry/>`_), and has been properly installed in your environment.

> You can check which ones you have currently available by running the ``verdi plugin`` command:

Reviewer: ```suggestion
You can check which ones you have currently available:
```

.. code-block:: bash

    $ verdi plugin list aiida.calculations

> For the purpose of this example, we will use the ``arithmetic.add`` plugin, which is a plugin for bash to sum two integers that comes pre-installed with ``aiida-core``.

Reviewer: ```suggestion
As an example, we will show how to use the ``arithmetic.add`` plugin, which is a pre-installed plugin implemented in `bash` to sum two integers.
```
The how-to is not an example, but it can utilize examples.

> You can access it through the use of the ``CalculationFactory``:

Reviewer: ```suggestion
You can access it with the ``CalculationFactory``:
```

.. code-block:: python

    from aiida.plugins import CalculationFactory
    calculation_class = CalculationFactory('arithmetic.add')

> We will obviously also need to provide the inputs for the code when running the calculation.

Reviewer: ```suggestion
Next, we provide the inputs for the code when running the calculation.
```

> You can also use ``verdi plugin`` to check this and other relevant information for specific plugins:

Reviewer: ```suggestion
Use ``verdi plugin`` to determine what inputs a specific plugin expects:
```

.. code-block:: bash

>     $ verdi plugin list aiida.calculations arithmetic.add

Reviewer: Please add the output here as well. We don't expect users to run these commands to be able to follow the example.

Author: Mmm, but from what I understood from @sphuber 's [previous review](https://github.com/aiidateam/aiida-core/pull/4114#discussion_r430023980), the idea was to not include the actual information in order for it to be easier to maintain. Maybe I misunderstood that?

Reviewer: That's not how I understand that comment and I was missing the output when I read through this how-to.

Reviewer 2: The idea was indeed to not have a copy of the spec here in case the implementation changes and the documentation goes out of date. Pasting the output of `verdi plugin list` has the same downside, but I can see why it is important to show it, so fine for me to add it. Maybe don't show the exit codes as they are not necessary and just do description, inputs and outputs.

Anyway, using this over copying source code is still better as they can then execute the command themselves to try and they can apply it to any other plugin, for example the one they would actually like to launch.

You will see that 3 inputs nodes are required: two containing the values to add up (``x``, ``y``) and one containing information about the specific code to execute (``code``).
If you already have these nodes in your database, you can get them by :ref:`querying for them <how-to:data:finding-data>` or using ``orm.load_node(<PK>)``.

> Otherwise, you will need to create them as shown below (note that you `will` need to already have the ``localhost`` computer configured):

Reviewer: We will need to provide a link where it is described how to do that or add a *dropdown* that shows this.

Author: You mean setting the localhost? Ok, yes, I agree. Would this be on the "Advanced Installation" section? I remember that when we were discussin the tutorial part of the assumptions was that they would have already had this set up, that's why I assumed it here.

Reviewer: I think it hasn't been merged yet. So feel free to just ignore this for now. I'll try to add meaningful links to the relevant section after everything has been merged.

Author: Should we add a reminder of this...somewhere?

Reviewer: I'll put it on the project board.

.. code-block:: python

    from aiida import orm
    code_node = orm.Code(remote_computer_exec=[localhost, '/bin/bash'])
    numx_node = orm.Int(17)
    numy_node = orm.Int(11)

To provide these as inputs to the calculations, we will now use the ``builder`` object that we can get from the class:

.. code-block:: python

    calculation_builder = calculation_class.get_builder()
    calculation_builder.code = code_node
    calculation_builder.x = numx_node
    calculation_builder.y = numy_node

> Now everything is in place and ready to perform the calculation, which can be done in two different ways.

Reviewer: I'm wondering if we can hand-off to a different section here. How to run and submit calculations is already described in a different how-to.

Author: Wait, what? Where? I thougth this was the how-to on running calculations.

Reviewer: Honestly, I might have gotten confused here. Let's leave this and I'll double check any potential duplication after everything is merged.

.. code-block:: python

    from aiida.engine import run
    output_dict = run(calculation_builder)
    sum_node = output_dict['sum']

The second one is non blocking, as you will be submitting it to the daemon and immediately getting control in the interpreter.
The return value in this case is the actual calculation node that is stored in the database.

.. code-block:: python

    from aiida.engine import submit
    calculation_node = submit(calculation_builder)

Note that, although you have access to the node, the underlying calculation `process` is not guaranteed to have finished when you get back control in the interpreter.
In order to keep track of it you can use the verdi command line interface:

.. code-block:: bash

    $ verdi process list

Performing a dry-run
--------------------

Additionally, you might want to check and verify your inputs before actually running or submitting a calculation.
You can do so by specifying to use a ``dry_run``, which will create all the input files in a local directory (``submit_test/[date]-0000[x]``) so you can inspect them before actually launching the calculation:

.. code-block:: python

    calculation_builder.metadata.dry_run = True

>     calculation_builder.metadata.store_provenance = False

Reviewer: Quick question: Does `dry_run = True` not imply `store_provenance = False`?

Author: @sphuber I do not know this, honestly I just used the same two lines that were present in the aiida tutorial of 2019.

Reviewer 2: No, dry run means the job does not actually get submitted and starts running, but simply creates the input files in a local folder. It will still create a node in the database and link the inputs up. If you do not want this (which is likely if you are running a dry run) you need to set it

Reviewer: Seems like "dry-run" is a bit of a misnomer then, but anyways, doesn't really affect the documentation.

Reviewer 2: why do you think it is a misnomer?

Reviewer: Because "dry-run" to me implies that nothing is actually going to happen, i.e., it is perfectly safe to execute without any effects. But I see that in this context it might make sense.

Running external codes
======================

To run an external code with AiiDA, you will need to use an appropriate :ref:`calculation plugin <topics:plugins>` that knows how to transform the input nodes into the input files that the code expects, copy everything in the code's machine, run the calculation and retrieve the results.

> This plugin would have already been created, either by you (see section on :ref:`interfacing external codes <how-to:codes:plugin>`) or by other AiiDA plugin-developers (available packages are listed in the `plugin registry <https://aiidateam.github.io/aiida-registry/>`_), and has been properly installed in your environment.

Reviewer 2: ```suggestion
You can check the `plugin registry <https://aiidateam.github.io/aiida-registry/>`_ to see if a plugin already exists for the code that you would like to run.
If that is not the case, you can :ref:`develop your own <how-to:codes:plugin>`.
After you have installed the plugin, you can start running the code through AiiDA.
```
Just wondering if we need a quick link for "installed the plugin" to a section on how to install plugins. Not sure if we already have it.

Author: Yeah, I thougth the same, but there is currently no such section. Should it be part of "How to manage your installation" maybe?

Reviewer 2: Yes make a section stub there called "Installing plugins" and create the corresponding issue, if you will please. Put it before the "Tuning performance" section, for now

> You can check which ones you have currently available:

Reviewer 2: ```suggestion
To check which calculation plugins you have currently installed, run:
```

.. code-block:: bash

    $ verdi plugin list aiida.calculations

> As an example, we will show how to use the ``arithmetic.add`` plugin, which is a pre-installed plugin implemented in `bash` to sum two integers.

Reviewer 2: ```suggestion
As an example, we will show how to use the ``arithmetic.add`` plugin, which is a pre-installed plugin that uses the `bash shell<https://www.gnu.org/software/bash/>`_ to sum two integers.
```

You can access it with the ``CalculationFactory``:

.. code-block:: python

    from aiida.plugins import CalculationFactory

>     calculation_class = CalculationFactory('arithmetic.add')

Reviewer: Just making sure, this is actually a class, right? And not an instance of a class, in which case it should be:
```suggestion
    calculation = CalculationFactory('arithmetic.add')
```

Author: Tis a class. Proof:
```
In [1]: from aiida.plugins import CalculationFactory
In [2]: import inspect
In [3]: calculation_class = CalculationFactory('arithmetic.add') 
In [4]: inspect.isclass(calculation_class)                                                                              
Out[4]: True
```

Next, we provide the inputs for the code when running the calculation.
Use ``verdi plugin`` to determine what inputs a specific plugin expects:

.. code-block:: bash

    $ verdi plugin list aiida.calculations arithmetic.add
    (...)
        Inputs:
               code:  required  Code        The `Code` to use for this job.
                  x:  required  Int, Float  The left operand.
                  y:  required  Int, Float  The right operand.
    (...)

You will see that 3 inputs nodes are required: two containing the values to add up (``x``, ``y``) and one containing information about the specific code to execute (``code``).
If you already have these nodes in your database, you can get them by :ref:`querying for them <how-to:data:finding-data>` or using ``orm.load_node(<PK>)``.

> Otherwise, you will need to create them as shown below (note that you `will` need to already have the ``localhost`` computer configured):

Reviewer 2: I think we really need at a minimum a link to a How-To configure a computer. I don't think it exists yet, but you can already create the link reference and stub, and open the corresponding issue

Author: Should it be part of "How to manage your installation"?

Reviewer 2: I would more be tempted to put it in this same How to, because really the only reason you need to configure a computer is to run `CalcJobs`. Not sure where, maybe after "Interfacing external codes" for now? Something like "Configuring a remote computer"

Author: ["Configuring remote resources"](https://github.com/aiidateam/aiida-core/issues/4123) sounds good?

Reviewer 2: I think "resources" is too broad in this case, because it does not necessarily convey that we are talking about computers. We could specify "Configuring remote computing resources" but in that case I think "Configuring remote computers" is better, as it is clearer and more concise, while saying the exact thing

Author: Ok, changed. Ready for final review now.

.. code-block:: python

    from aiida import orm
    code_node = orm.Code(remote_computer_exec=[localhost, '/bin/bash'])
    numx_node = orm.Int(17)

>     numy_node = orm.Int(11)

Reviewer: In other sections we have not used the suffix `_node` so far. I'd prefer to drop that everywhere.

To provide these as inputs to the calculations, we will now use the ``builder`` object that we can get from the class:

.. code-block:: python

    calculation_builder = calculation_class.get_builder()
    calculation_builder.code = code_node
    calculation_builder.x = numx_node
    calculation_builder.y = numy_node

Now everything is in place and ready to perform the calculation, which can be done in two different ways.
The first one is blocking and will return a dictionary containing all the output nodes (keyed after their label, so in this case these should be: "remote_folder", "retrieved" and "sum") that you can safely inspect and work with:

.. code-block:: python

    from aiida.engine import run
    output_dict = run(calculation_builder)
    sum_node = output_dict['sum']

> The second one is non blocking, as you will be submitting it to the daemon and immediately getting control in the interpreter.

Reviewer: ```suggestion
The second one is non blocking, as you will be submitting it to the daemon and control is immediately returned to the interpreter.
```

> The return value in this case is the actual calculation node that is stored in the database.

Reviewer: ```suggestion
The return value in this case is the calculation node that is stored in the database.
```
I'm not sure what we try to convey here with "actual".

.. code-block:: python

    from aiida.engine import submit
    calculation = submit(calculation_builder)

Note that, although you have access to the node, the underlying calculation `process` is not guaranteed to have finished when you get back control in the interpreter.
In order to keep track of it you can use the verdi command line interface:

.. code-block:: bash

>     $ verdi process list

Reviewer 2: Maybe a link here to the topic on monitoring processes: https://aiida.readthedocs.io/projects/aiida-core/en/docs-revamp/topics/processes/usage.html#monitoring-processes

Running external codes
======================

> The process of running an external code through AiiDA involves loading and instantiating the appropriate ``CalcJob`` wrapping class for the code, filling the inputs necessary, and then running or submitting the constructed object.

Reviewer 2: "The process of running an external code .... involves" is pretty verbose language for a how-to.
How about something more direct and to the point:
> To run an external code with AiiDA, you need a :ref:`plugin <correct link to more info>` for it.
> The plugin's job is to transform the input *nodes* to the input files that the code expects.
> If the plugin for your code already exists (see the `plugin registry <http://registry>`_ what is available) and you have installed the :ref:`plugin package <ref>`, you can load it using the ``CalculationFactory``.

If the ``CalcJob`` has been registered in AiiDA (see section :ref:`on how to register plugins<how-to:plugins:entrypoints>`) you can use the ``CalculationFactory``:

.. code-block:: python

>     from aiida.plugins import CalculationFactory
>     calculation_builder = CalculationFactory('arithmetic.add').get_builder()

Reviewer 2: ```suggestion
    from aiida.plugins import CalculationFactory
    calculation_class = CalculationFactory('arithmetic.add')
```
I would assign the class explicitly to make it clear that that is what the factory returns.
I would also not use the builder straight away but go for the simple dictionary syntax

Otherwise, you can just load the class and get the builder directly:

.. code-block:: python

    from aiida.calculations.plugins.arithmetic.add import ArithmeticAddCalculation

>     calculation_builder = ArithmeticAddCalculation.get_builder()

Reviewer 2: This will very often lead to problems when they start submitting to the daemon and I would not mention it here as it is an unnecessary alternative.

> Then you need to start filling out the builder with all the inputs that you need to specify to run your calculation.

Reviewer 2: Should add a line here saying:
> For the purpose of this example, we will use the `arithmetic.add` plugin, which is a plugin for `bash` to sum two integers

In the case of the ``ArithmeticAddCalculation``, there are 4 required inputs:

.. code-block:: python
    :linenos:

    # Specific for ArithmeticAdd
    spec.input('x', valid_type=(orm.Int, orm.Float), help='The left operand.')
    spec.input('y', valid_type=(orm.Int, orm.Float), help='The right operand.')

    # General for all CalcJobs
    spec.input('code', valid_type=orm.Code, help='The `Code` to use for this job.')
    spec.input('metadata.options.resources', valid_type=dict, required=True, validator=validate_resources,
        help='Set the dictionary of resources to be used by the scheduler plugin, like the number of nodes, '
             'cpus etc. This dictionary is scheduler-plugin dependent. Look at the documentation of the '

>              'scheduler for more details.')

Reviewer 2: Instead of this I would suggest to use `verdi plugin list aiida.calculations arithmetic.add`
This way we don't have to worry about updating this when it changes, and it will also apply to any other calculation plugin

Reviewer 2: Would also remove the mention of resources, which will be simplified by Marnik's changes that will soon be merged in `develop`.

Author: Ok, so, the drawback of removing the `resources` is that in this example only nodes would be provided, which might make people think you can only pass nodes to the calculations. Are we ok with this? Should I maybe add a change to one of the optional parameters (like the `output_filename`)?

Reviewer 2: I think that is actually ok for this introductory how-to. We could make another on in the future, something like "How to change scheduler specific inputs", which will then provide a natural way to introduce `metadata` inputs.

The first three are input nodes (lines 2, 3 and 6), so you can either use nodes that are already in your database (you can query for them, load them, etc.) or you can create new nodes and these will be automatically stored when starting the calculation.

.. code-block:: python

    code_node = orm.load_node(100)
    num1_node = orm.Int(17)
    num2_node = orm.Int(11)

    calculation_builder.code = code_node
    calculation_builder.x = num1_node
    calculation_builder.y = num2_node

> The last input (line 7) is not a node but a dictionary containing information about the resources required from the remote machine to run the code.

Reviewer 2: This line is incorrect I think. You are not adding the metadata here, but in the next block

Author: I guess it is no longer relevant if I'm removing the metadata, but the line "The last input (line 7)" refers not to the last code block, but the previous one. I was hoping it would have been a bit more clear looking at the compiled doc because it was the only code block with numbers in the section, but it is true that it is not ideally phrased.

It will be stored as a property of the ``CalcJob`` node, and should at least contain the following two keys:

.. code-block:: python

    calculation_builder.metadata.options.resources = {
        'num_machines': 1,
        'tot_num_mpiprocs': 1,

>     }

Reviewer 2: as said before, I think we should remove this. These are not the correct settings for all schedulers and soon this will be set automatically. This will simplify it a lot.

Now everything is in place and you are ready to perform the calculation, which you can do in two different ways.
The first one is blocking and will return a dictionary containing all the output nodes (keyed after their label, so in this case these should be: "remote_folder", "retrieved" and "sum") that you can safely inspect and work with:

.. code-block:: python

    from aiida.engine import run
    output_dict = run(calculation_builder)
    sum_node = output_dict['sum']

The second one is non blocking, as you will be submitting it to the daemon and immediately getting control in the interpreter:

.. code-block:: python

    from aiida.engine import submit
    calculation_node = submit(calculation_builder)

As the variable name suggests, the return value for the submit method is the calculation node that is stored in the database.
In this case the underlying calculation is not guaranteed to have finished, so it is often convenient to keep track of it using the verdi command line interface:

.. code-block:: bash

    $ verdi process list

> 

Reviewer 2: Should probably create a subsection for "Performing a dry-run"

Additionally, you might want to check and verify your inputs before actually running or submitting a calculation.

> You can do so by specifying to use a ``dry_run``, which will create the running folder in the current directory (``submit_test/[date]-0000[x]``) so the user can inspect the integrity of the inputs generated:

Reviewer 2: ```suggestion
You can do so by specifying to use a ``dry_run``, which will create all the input files in a local directory (``submit_test/[date]-0000[x]``) so you can inspect them before actually launching the calculation:
```
